{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Algorithms.Z_algorithm import get_matches as Z\n",
    "from Algorithms.manacher import manacher as M\n",
    "\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clase principal\n",
    "class Virus:\n",
    "  def __init__(self, file_path):\n",
    "    if not os.path.exists(file_path):\n",
    "      raise FileNotFoundError(f\"El archivo {file_path} no existe\")\n",
    "    with open(file_path, 'r') as file:\n",
    "      lines = file.readlines()\n",
    "    self.metadata = lines[0].strip()\n",
    "    self.sequence = ''.join(line.strip() for line in lines[1:])\n",
    "    self.file = os.path.basename(file_path)\n",
    "    self.palindromo = M(self.sequence) # Algoritmo Manacher importado\n",
    "    \n",
    "    new_sequence = ''\n",
    "    # Iterar por la longitud de la secuencia con un salto de 3 (cada codon)\n",
    "    for i in range(0, len(self.sequence), 3):  \n",
    "        codon = self.sequence[i:i+3]\n",
    "        amino_acid = self.codon_to_amino_acid.get(codon, '?')  # '?' si no encuentra\n",
    "        new_sequence += amino_acid\n",
    "    self.amino_sequence = new_sequence\n",
    "\n",
    "  def getFirstChars(self, n=12):\n",
    "    if n > len(self.sequence):\n",
    "      raise ValueError(\"n no puede ser mayor a la longitud de la secuencia (length =\", len(self.sequence), \")\")\n",
    "    return self.sequence[:n]\n",
    "  \n",
    "  codon_to_amino_acid = {\n",
    "    'GCT': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A',  # Alanine\n",
    "    'CGT': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R', 'AGA': 'R', 'AGG': 'R',  # Arginine\n",
    "    'AAT': 'N', 'AAC': 'N',  # Asparagine\n",
    "    'GAT': 'D', 'GAC': 'D',  # Aspartic Acid\n",
    "    'TGT': 'C', 'TGC': 'C',  # Cysteine\n",
    "    'CAA': 'Q', 'CAG': 'Q',  # Glutamine\n",
    "    'GAA': 'E', 'GAG': 'E',  # Glutamic Acid\n",
    "    'GGT': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G',  # Glycine\n",
    "    'CAT': 'H', 'CAC': 'H',  # Histidine\n",
    "    'ATT': 'I', 'ATC': 'I', 'ATA': 'I',  # Isoleucine\n",
    "    'TTA': 'L', 'TTG': 'L', 'CTT': 'L', 'CTC': 'L', 'CTA': 'L', 'CTG': 'L',  # Leucine\n",
    "    'AAA': 'K', 'AAG': 'K',  # Lysine\n",
    "    'ATG': 'M',  # Methionine (Start)\n",
    "    'TTT': 'F', 'TTC': 'F',  # Phenylalanine\n",
    "    'CCT': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P',  # Proline\n",
    "    'TCT': 'S', 'TCC': 'S', 'TCA': 'S', 'TCG': 'S', 'AGT': 'S', 'AGC': 'S',  # Serine\n",
    "    'ACT': 'T', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T',  # Threonine\n",
    "    'TGG': 'W',  # Tryptophan\n",
    "    'TAT': 'Y', 'TAC': 'Y',  # Tyrosine\n",
    "    'GTT': 'V', 'GTC': 'V', 'GTA': 'V', 'GTG': 'V',  # Valine\n",
    "    'TAA': '*', 'TAG': '*', 'TGA': '*'  # Stop \n",
    "  } \n",
    "\n",
    "  def printData(self):\n",
    "    print(\"File: \", self.file)\n",
    "    print(\"Metadata: \", self.metadata)\n",
    "    #print(\"Sequence: \", self.sequence)\n",
    "    print(\"Primeros 12 chars: \", self.getFirstChars())\n",
    "    print(\"Palindromo: \", self.palindromo)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Gen(Virus):\n",
    "  def __init__(self, file_path):\n",
    "    super().__init__(file_path)\n",
    "    self.VirusIndexes = []\n",
    "\n",
    "  def check_viruses(self, viruses):\n",
    "    for virus in viruses:\n",
    "      self.VirusIndexes.append([virus.file, Z(self.sequence, virus.sequence)])\n",
    "\n",
    "  def printData(self):\n",
    "    super().printData()\n",
    "    for virus in self.VirusIndexes:\n",
    "      print(\"Virus: \", virus[0])\n",
    "      print(\"Indexes: \", virus[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Proteina(Virus):\n",
    "  def __init__(self, metadata, sequence, file_path):\n",
    "    self.file = os.path.basename(file_path)\n",
    "    self.metadata = metadata\n",
    "    self.sequence = sequence\n",
    "    self.palindromo = M(self.sequence)  # Manacher function or similar\n",
    "    self.VirusIndexes = []\n",
    "    self.GeneIndexes = []\n",
    "\n",
    "  @classmethod\n",
    "  def leer_de_mismo_archivo(cls, file_path):\n",
    "    if not os.path.exists(file_path):\n",
    "      raise FileNotFoundError(f\"El archivo {file_path} no existe\")\n",
    "\n",
    "    with open(file_path, 'r') as file:\n",
    "      content = file.read().split('>')  # Split by '>' to separate different proteins\n",
    "\n",
    "      proteins = []\n",
    "      for entry in content:\n",
    "        if entry.strip():\n",
    "          lines = entry.splitlines()\n",
    "          metadata = lines[0].strip()\n",
    "          sequence = ''.join(line.strip() for line in lines[1:])\n",
    "          protein = cls(metadata, sequence, file_path)  # Create Proteina instances\n",
    "          proteins.append(protein)\n",
    "\n",
    "      return proteins\n",
    "  \n",
    "  # Sobreescribir porque ahora se busca con el amino_sequence\n",
    "  def check_viruses(self, viruses):\n",
    "    for virus in viruses:\n",
    "      self.VirusIndexes.append([virus.file, Z(self.sequence, virus.amino_sequence)])\n",
    "  \n",
    "  def check_genes(self, genes):\n",
    "    for gene in genes:\n",
    "      self.GeneIndexes.append([gene.file, Z(self.sequence, gene.amino_sequence)])\n",
    "\n",
    "  def printData(self):\n",
    "    print(\"Metadata: \", self.metadata)\n",
    "    #print(\"Sequence: \", self.sequence)\n",
    "    print(\"Primeros 12 chars: \", self.getFirstChars())\n",
    "    print(\"Palindromo: \", self.palindromo)\n",
    "    for virus in self.VirusIndexes:\n",
    "      print(\"Virus: \", virus[0])\n",
    "      print(\"Indexes: \", virus[1])\n",
    "    for gene in self.GeneIndexes:\n",
    "      print(\"Gene: \", gene[0])\n",
    "      print(\"Indexes: \", gene[1])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def leer_archivos_de_directorio(directory, tipo):\n",
    "    sequences = []\n",
    "    for filename in os.listdir(directory):\n",
    "      if filename.endswith('.txt'):\n",
    "        file_path = os.path.join(directory, filename)\n",
    "        if tipo == 1:\n",
    "          sequences.append(Virus(file_path))\n",
    "        elif tipo == 2:\n",
    "          sequences.append(Gen(file_path))\n",
    "        elif tipo == 3:\n",
    "          # Aqui es extend porque regresa una lista de objetos y no un objeto (y se rquiere una lista de objetos plana)\n",
    "          sequences.extend(Proteina.leer_de_mismo_archivo(file_path))\n",
    "        else:\n",
    "          raise ValueError(\"Tipo no valido\")\n",
    "    return sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----------------------\n",
      "\n",
      "VIRUS DATA\n",
      "\n",
      "-----------------------\n",
      "\n",
      "File:  SARS-COV-2-MN908947.3.txt\n",
      "Metadata:  >MN908947.3 Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome\n",
      "Primeros 12 chars:  ATTAAAGGTTTA\n",
      "Palindromo:  AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      "\n",
      "File:  SARS-COV-2-MT106054.1.txt\n",
      "Metadata:  >MT106054.1 Severe acute respiratory syndrome coronavirus 2 isolate SARS-CoV-2/human/USA/TX-CDC-03050000-001/2020, complete genome\n",
      "Primeros 12 chars:  ATTAAAGGTTTA\n",
      "Palindromo:  CTCAATGACTTCAGTAACTC\n",
      "\n",
      "-----------------------\n",
      "\n",
      "GENE DATA\n",
      "\n",
      "-----------------------\n",
      "\n",
      "File:  gen-M.txt\n",
      "Metadata:  >MN908947.3:26523-27191 Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome\n",
      "Primeros 12 chars:  ATGGCAGATTCC\n",
      "Palindromo:  CTAAAGAAATC\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  [26522]\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  [26522]\n",
      "\n",
      "File:  gen-ORF1AB.txt\n",
      "Metadata:  >MN908947.3:266-21555 Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome\n",
      "Primeros 12 chars:  ATGGAGAGCCTT\n",
      "Palindromo:  CTCAATGACTTCAGTAACTC\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  [265]\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  []\n",
      "\n",
      "File:  gen-S.txt\n",
      "Metadata:  >MN908947.3:21563-25384 Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome\n",
      "Primeros 12 chars:  ATGTTTGTTTTT\n",
      "Palindromo:  ATTTTGGTGGTTTTA\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  [21562]\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  [21562]\n",
      "\n",
      "-----------------------\n",
      "\n",
      "PROTEIN DATA\n",
      "\n",
      "-----------------------\n",
      "\n",
      "Metadata:  QHD43415_1\n",
      "Primeros 12 chars:  MESLVPGFNEKT\n",
      "Palindromo:  KNGNK\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  []\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  [0]\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  []\n",
      "\n",
      "Metadata:  QHD43415_2\n",
      "Primeros 12 chars:  AYTRYVDNNFCG\n",
      "Palindromo:  AKLKA\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  []\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  [180]\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  []\n",
      "\n",
      "Metadata:  QHD43415_3\n",
      "Primeros 12 chars:  APTKVTFGDDTV\n",
      "Palindromo:  VNAANV\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  []\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  [818]\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  []\n",
      "\n",
      "Metadata:  QHD43415_4\n",
      "Primeros 12 chars:  KIVNNWLKQLIK\n",
      "Palindromo:  LAKAL\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  []\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  [2763]\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  []\n",
      "\n",
      "Metadata:  QHD43415_5\n",
      "Primeros 12 chars:  SGFRKMAFPSGK\n",
      "Palindromo:  GSPSG\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  []\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  [3263]\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  []\n",
      "\n",
      "Metadata:  QHD43415_6\n",
      "Primeros 12 chars:  SAVKRTIKGTHH\n",
      "Palindromo:  VLTLV\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  []\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  [3569]\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  []\n",
      "\n",
      "Metadata:  QHD43415_7\n",
      "Primeros 12 chars:  SKMSDVKCTSVV\n",
      "Palindromo:  VSLLSV\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  []\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  [3859]\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  []\n",
      "\n",
      "Metadata:  QHD43415_8\n",
      "Primeros 12 chars:  AIASEFSSLPSY\n",
      "Palindromo:  NNIINN\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  []\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  [3942]\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  []\n",
      "\n",
      "Metadata:  QHD43415_9\n",
      "Primeros 12 chars:  NNELSPVALRQM\n",
      "Palindromo:  LNNL\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  []\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  [4140]\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  []\n",
      "\n",
      "Metadata:  QHD43415_10\n",
      "Primeros 12 chars:  AGNATEVPANST\n",
      "Palindromo:  CSC\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  []\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  [4253]\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  []\n",
      "\n",
      "Metadata:  QHD43415_11\n",
      "Primeros 12 chars:  SADAQSFLNRVC\n",
      "Palindromo:  TNDNT\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  []\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  []\n",
      "\n",
      "Metadata:  QHD43415_12\n",
      "Primeros 12 chars:  AVGACVLCNSQT\n",
      "Palindromo:  ARLRA\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  [5412]\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  [5412]\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  []\n",
      "\n",
      "Metadata:  QHD43415_13\n",
      "Primeros 12 chars:  AENVTGLFKDCS\n",
      "Palindromo:  TKFKT\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  [6013]\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  []\n",
      "\n",
      "Metadata:  QHD43415_14\n",
      "Primeros 12 chars:  SLENVAFNVVNK\n",
      "Palindromo:  DLLLD\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  [6540]\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  [6540]\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  []\n",
      "\n",
      "Metadata:  QHD43415_15\n",
      "Primeros 12 chars:  SSQAWQPGVAMP\n",
      "Palindromo:  SDADS\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  [6886]\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  [6886]\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  []\n",
      "\n",
      "Metadata:  QHD43416\n",
      "Primeros 12 chars:  MFVFLVLLPLVS\n",
      "Palindromo:  ADTTDA\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  []\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  [0]\n",
      "\n",
      "Metadata:  QHD43417\n",
      "Primeros 12 chars:  MDLFMRIFTIGT\n",
      "Palindromo:  VALLAV\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  [8464]\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  [8464]\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  []\n",
      "\n",
      "Metadata:  QHD43418\n",
      "Primeros 12 chars:  MYSFVSEETGTL\n",
      "Palindromo:  FVVF\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  [8748]\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  [8748]\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  []\n",
      "\n",
      "Metadata:  QHD43419\n",
      "Primeros 12 chars:  MADSNGTITVEE\n",
      "Palindromo:  LESEL\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  []\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  [0]\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  []\n",
      "\n",
      "Metadata:  QHD43420\n",
      "Primeros 12 chars:  MFHLVDFQVTIA\n",
      "Palindromo:  LSKSL\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  [9067]\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  [9067]\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  []\n",
      "\n",
      "Metadata:  QHD43421\n",
      "Primeros 12 chars:  MKIILFLALITL\n",
      "Palindromo:  KRK\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  [9131]\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  [9131]\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  []\n",
      "\n",
      "Metadata:  QHD43422\n",
      "Primeros 12 chars:  MKFLVFLGIITT\n",
      "Palindromo:  VRV\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  []\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  []\n",
      "\n",
      "Metadata:  QHD43423\n",
      "Primeros 12 chars:  MSDNGPQNQRNA\n",
      "Palindromo:  SRSSSRS\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  []\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  []\n",
      "\n",
      "Metadata:  QHI42199\n",
      "Primeros 12 chars:  MGYINVFAFPFT\n",
      "Palindromo:  NFN\n",
      "Virus:  SARS-COV-2-MN908947.3.txt\n",
      "Indexes:  []\n",
      "Virus:  SARS-COV-2-MT106054.1.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-M.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-ORF1AB.txt\n",
      "Indexes:  []\n",
      "Gene:  gen-S.txt\n",
      "Indexes:  []\n",
      "\n"
     ]
    }
   ],
   "source": [
    "path_genes = 'genes'\n",
    "path_viruses = 'viruses'\n",
    "path_proteinas = 'proteinas'\n",
    "viruses = leer_archivos_de_directorio(path_viruses, 1)\n",
    "genes = leer_archivos_de_directorio(path_genes, 2)\n",
    "proteinas = leer_archivos_de_directorio(path_proteinas, 3)\n",
    "\n",
    "print(\"-----------------------\")\n",
    "print(\"\\nVIRUS DATA\\n\")\n",
    "print(\"-----------------------\\n\")\n",
    "for virus in viruses:\n",
    "    virus.printData()\n",
    "    print()\n",
    "\n",
    "print(\"-----------------------\")\n",
    "print(\"\\nGENE DATA\\n\")\n",
    "print(\"-----------------------\\n\")\n",
    "\n",
    "for gen in genes:\n",
    "    gen.check_viruses(viruses)\n",
    "    gen.printData()\n",
    "    print()\n",
    "\n",
    "print(\"-----------------------\")\n",
    "print(\"\\nPROTEIN DATA\\n\")\n",
    "print(\"-----------------------\\n\")\n",
    "\n",
    "for proteina in proteinas:\n",
    "    proteina.check_viruses(viruses)\n",
    "    proteina.check_genes(genes)\n",
    "    proteina.printData()\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'SARS-COV-2-MN908947.3.txt': [('gen-M.txt', [26522]), ('gen-ORF1AB.txt', [265]), ('gen-S.txt', [21562])], 'SARS-COV-2-MT106054.1.txt': [('gen-M.txt', [26522]), ('gen-S.txt', [21562])]}\n",
      "{'gen-M.txt': [('QHD43419', [0])], 'gen-ORF1AB.txt': [('QHD43415_1', [0]), ('QHD43415_2', [180]), ('QHD43415_3', [818]), ('QHD43415_4', [2763]), ('QHD43415_5', [3263]), ('QHD43415_6', [3569]), ('QHD43415_7', [3859]), ('QHD43415_8', [3942]), ('QHD43415_9', [4140]), ('QHD43415_10', [4253])], 'gen-S.txt': [('QHD43416', [0])]}\n"
     ]
    }
   ],
   "source": [
    "virus_map = {}\n",
    "virusIndex = 0\n",
    "for virus in viruses:\n",
    "    virus_map[virus.file] = []\n",
    "    for gene in genes:\n",
    "        if len(gene.VirusIndexes[virusIndex][1]) > 0:\n",
    "            virus_map[virus.file].append((gene.file, gene.VirusIndexes[virusIndex][1]))\n",
    "    virusIndex += 1\n",
    "print(virus_map)\n",
    "\n",
    "gene_map = {}\n",
    "geneIndex = 0\n",
    "for gene in genes:\n",
    "    gene_map[gene.file] = []\n",
    "    for proteina in proteinas:\n",
    "        if len(proteina.GeneIndexes[geneIndex][1]) > 0:\n",
    "            gene_map[gene.file].append((proteina.metadata, proteina.GeneIndexes[geneIndex][1]))\n",
    "    geneIndex += 1\n",
    "print(gene_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapeo de virus a gen:\n",
      "\n",
      "SARS-COV-2-MN908947.3.txt:\n",
      "  - gen-M.txt en indice [26522]\n",
      "  - gen-ORF1AB.txt en indice [265]\n",
      "  - gen-S.txt en indice [21562]\n",
      "\n",
      "SARS-COV-2-MT106054.1.txt:\n",
      "  - gen-M.txt en indice [26522]\n",
      "  - gen-S.txt en indice [21562]\n",
      "\n",
      "Mapeo de proteina a gen:\n",
      "\n",
      "gen-M.txt:\n",
      "  - Proteina QHD43419 en indice [0]\n",
      "\n",
      "gen-ORF1AB.txt:\n",
      "  - Proteina QHD43415_1 en indice [0]\n",
      "  - Proteina QHD43415_2 en indice [180]\n",
      "  - Proteina QHD43415_3 en indice [818]\n",
      "  - Proteina QHD43415_4 en indice [2763]\n",
      "  - Proteina QHD43415_5 en indice [3263]\n",
      "  - Proteina QHD43415_6 en indice [3569]\n",
      "  - Proteina QHD43415_7 en indice [3859]\n",
      "  - Proteina QHD43415_8 en indice [3942]\n",
      "  - Proteina QHD43415_9 en indice [4140]\n",
      "  - Proteina QHD43415_10 en indice [4253]\n",
      "\n",
      "gen-S.txt:\n",
      "  - Proteina QHD43416 en indice [0]\n"
     ]
    }
   ],
   "source": [
    "# Print virus map with better formatting\n",
    "print(\"Mapeo de virus a gen:\")\n",
    "for virus, genes in virus_map.items():\n",
    "    print(f\"\\n{virus}:\")\n",
    "    for gene, index in genes:\n",
    "        print(f\"  - {gene} en indice {index}\")\n",
    "\n",
    "# Print gene map with better formatting\n",
    "print(\"\\nMapeo de proteina a gen:\")\n",
    "for gene, proteins in gene_map.items():\n",
    "    print(f\"\\n{gene}:\")\n",
    "    for protein, index in proteins:\n",
    "        print(f\"  - Proteina {protein} en indice {index}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
